test/fx/test_subgraph_rewriter.py

```diff
@@ -0,0 +1,188 @@
+import os
+import sys
+
+import torch
+from torch.fx import symbolic_trace, subgraph_rewriter
+
+# Make the helper files in test/ importable
+pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+sys.path.append(pytorch_test_dir)
+from torch.testing._internal.jit_utils import JitTestCase
+
+if __name__ == '__main__':
+    raise RuntimeError("This test file is not meant to be run directly, use:\n\n"
+                       "\tpython test/test_fx.py TESTNAME\n\n"
+                       "instead.")
+
+class TestSubgraphRewriter(JitTestCase):
+
+    def test_subgraph_rewriter_preserves_logic(self):
+        # Function that we want to perform the rewrite on
+        def fn(x, w1, w2, b1, b2):
+            m1 = torch.cat([w1, w2])
+            m2 = torch.cat([x, b2])
+            t1 = torch.sum(w1, 1)
+            t2 = torch.addmm(b1, m1, m2.t())
+            return torch.eq(torch.sum(t1), torch.sum(t2))
+
+        # Pattern we want to match in `fn`
+        def pattern(x, w1, w2, b1, b2):
+            m1 = torch.cat([w1, w2])
+            m2 = torch.cat([x, b2])
+            return torch.addmm(b1, m1, m2.t())
+
+        # Set up
+        traced_fn = symbolic_trace(fn)
+        comparison_fn = symbolic_trace(fn)
+
+        # Auxiliary testing variables
+        x = torch.rand(1, 3)
+        w1 = torch.rand(1, 3)
+        w2 = torch.rand(1, 3)
+        b1 = torch.rand(2, 2)
+        b2 = torch.rand(1, 3)
+
+        # Replace `pattern` with the same pattern
+        subgraph_rewriter.replace_pattern(traced_fn.graph, pattern, pattern)
```

Comment:
Mutating the `graph` of a `GraphModule` in-place requires you to call `Graph.recompile()` afterward for the `GraphModule` code to be updated. I think there are a few options here in terms of API design:

* Keep the `replace_pattern` API as-is and require the user to manuall call `recompile()`. I'm worried this will be error-prone, however.
* Make it so that `replace_pattern` returns a new `Graph` instead of mutating the passed-in graph. Then the user would have to explicitly assign that graph into the GraphModule, forcing a recompile
* Make it so that `replace_pattern` takes in a `GraphModule` and handles recompilation manually. This might be the best course of action long-term, since we may need to add functionality to the subgraph rewriter to match on and replace `call_module` nodes, and that may involve mutating the `Module` hierarchy (not sure what the API for that should exactly be but it's a possibility)

Reply:
FWIW after calling `recompile()` there's some issues. The code for before and after has a diff:

```
@@ -1,21 +1,18 @@
 import torch
-def forward(self, x, w1, w2, b1, b2):
+def forward(self, b2, b1, w2, w1, x):
+    sum_1 = torch.sum(w1, 1)
+    sum_2 = torch.sum(sum_1)
+    sum_1 = None
     cat_1 = torch.cat([w1, w2])
-    w2 = None
+    w1 = w2 = None
     cat_2 = torch.cat([x, b2])
     x = b2 = None
-    sum_1 = torch.sum(w1, 1)
-    w1 = None
     t_1 = cat_2.t()
     cat_2 = None
-    addmm_1 = torch.addmm(b1, cat_1, t_1)
+    addmm_2 = torch.addmm(b1, cat_1, t_1)
     b1 = cat_1 = t_1 = None
-    sum_2 = torch.sum(sum_1)
-    sum_1 = None
-    sum_3 = torch.sum(addmm_1)
-    addmm_1 = None
-    eq_1 = torch.eq(sum_2, sum_3)
-    sum_2 = sum_3 = None
+    eq_1 = torch.eq(sum_2, addmm_2)
+    sum_2 = addmm_2 = None
     return eq_1
     eq_1 = None
```

And the invocation fails at `cat_1`. I'll see if I can dig in and see what's going on

Reply:
OK, what's going on here is that placeholders from the pattern graph are being inserted in reverse order, which changes how arguments appear positionally in the input graph.

As mentioned somewhere down there, I actually think that `placeholders` should be handled differently, i.e. the placeholders from the pattern graph should not be inserted into the input graph, but rather those are notional and refer to arbitrary values that are used in the pattern